libraries/aws/shadow/utest/shadow_utest.c

```diff
@@ -0,0 +1,494 @@
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_STRING_DELETE( TEST_THING_NAME ), TEST_SHADOW_TOPIC_STRING_DELETE );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_STRING_DELETE_ACCEPTED( TEST_THING_NAME ), TEST_SHADOW_TOPIC_STRING_DELETE_ACCEPTED );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_STRING_DELETE_REJECTED( TEST_THING_NAME ), TEST_SHADOW_TOPIC_STRING_DELETE_REJECTED );
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Tests the macros generates the expected strings length.
+ */
+void test_Shadow_MacrosLength( void )
+{
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_UPDATE( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_UPDATE );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_UPDATE_ACCEPTED( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_UPDATE_ACCEPTED );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_UPDATE_REJECTED( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_UPDATE_REJECTED );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_UPDATE_DOCUMENTS( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_UPDATE_DOCUMENTS );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_UPDATE_DELTA( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_UPDATE_DELTA );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_GET( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_GET );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_GET_ACCEPTED( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_GET_ACCEPTED );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_GET_REJECTED( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_GET_REJECTED );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_DELETE( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_DELETE );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_DELETE_ACCEPTED( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_DELETE_ACCEPTED );
+    TEST_ASSERT_EQUAL( SHADOW_TOPIC_LENGTH_DELETE_REJECTED( TEST_THING_NAME_LENGTH ), TEST_SHADOW_TOPIC_LENGTH_DELETE_REJECTED );
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Tests the behavior of Shadow_GetTopicString() with an invalid
+ *        document info parameter.
+ */
+void test_Shaodw_GetTopicString_Invalid_Parameters( void )
+{
+    ShadowStatus_t shadowStatus = SHADOW_STATUS_SUCCESS;
+    uint16_t outLength = 0;
+    ShadowTopicStringType_t topicType = SHADOW_TOPIC_STRING_TYPE_GET_ACCEPTED;
+    char topicBuffer[ TEST_SHADOW_TOPIC_BUFFER_LENGTH ] = TEST_SHADOW_TOPIC_BUFFER_INITIALZIE;
+    uint16_t bufferSize = TEST_SHADOW_TOPIC_BUFFER_LENGTH;
+    char topicBufferGetAccepted[ SHADOW_TOPIC_LENGTH_GET_ACCEPTED( TEST_THING_NAME_LENGTH ) ] = { 0 };
+    uint16_t bufferSizeGetAccepted = TEST_SHADOW_TOPIC_LENGTH_GET_ACCEPTED;
+
+    /* Call Shadow_GetTopicString() with various combinations of
+     * incorrect paramters. */
+    shadowStatus = Shadow_GetTopicString( 0,
+                                          "",
+                                          0,
+                                          & ( topicBuffer[ 0 ] ),
```

Comment:
I think we already have that test cases in line #350 - #371 

    shadowStatus = Shadow_GetTopicString( topicType,
                                          TEST_THING_NAME,
                                          TEST_THING_NAME_LENGTH,
                                          & ( topicBuffer[ 0 ] ),
                                          ( bufferSize / 2 ),
                                          & outLength );
    TEST_ASSERT_EQUAL_INT( SHADOW_STATUS_BUFFER_TOO_SMALL, shadowStatus );

    /* Call Shadow_GetTopicString() with valid paramters but bufferSize > topic string length
     * and verify result. */
    shadowStatus = Shadow_GetTopicString( topicType,
                                          TEST_THING_NAME,
                                          TEST_THING_NAME_LENGTH,
                                          & ( topicBuffer[ 0 ] ),
                                          bufferSize,
                                          & outLength );
    TEST_ASSERT_EQUAL_INT( SHADOW_STATUS_SUCCESS, shadowStatus );
    TEST_ASSERT_EQUAL_INT( TEST_SHADOW_TOPIC_LENGTH_GET_ACCEPTED, outLength );
    TEST_ASSERT_LESS_THAN( bufferSize, outLength );
    TEST_ASSERT_EQUAL_STRING_LEN( TEST_SHADOW_TOPIC_BUFFER_MODIFIED,
                                  topicBuffer,
                                  bufferSize );


not sure if this the case you are looking for:
( bufferSize / 2 ) will have less buffer size then the generated string length

TEST_SHADOW_TOPIC_BUFFER_MODIFIED is the test case for extra padding

Please let me know if we still need to add a test case

Thank you

Reply:
You're right, it does test the padding at the end, I was wondering if we should add an offset to test the padding at the beginning as well. I started thinking of that since I thought it looked a little odd to use `& ( topicBuffer[ 0 ] )` instead of just `topicBuffer` if there isn't going to be an offset. Just a minor point though so I don't think it needs to be changed

Reply:
The check Muneeb mentions is to check that the API does not write out of bounds in the happy case. In other words, if the memory provided to the API was from x to y, check that it does not write anything before x or after y. Few padding bytes (0xA5) are placed before x and after y prior to the API call and are verified to be unchanged after the API call.

```
	                        Memory passed to the API	
                    <-------------------------------------------->
        +----+----+----+----+----+----+----+----+----+----+----+----+----+----+
        |    |    |    |    |    |    |    |    |    |    |    |    |    |    |
        |0xA5|0xA5|    |    |    |    |    |    |    |    |    |    |0xA5|0xA5|
        +----+----+----+----+----+----+----+----+----+----+----+----+----+----+
                    ^                                            ^
                    |                                            |
                    |                                            |
                    x                                            y
```

Reply:
>  I thought it looked a little odd to use & ( topicBuffer[ 0 ] ) instead of just topicBuffer if there isn't going to be an offset. Just a minor point though so I don't think it needs to be changed

I have seen some compilers breaking with the `topicBuffer`, therefore, I suggest to always use `& ( topicBuffer[ 0 ] )`.